### PR TITLE
Pass the compile options for deserialization via PJRT C API

### LINKED ISSUE #18286: Deserialization of executables fails on non-zero ranks when deserializing single-device executable
If we deserialize executables via the wrapper C API client, [the compile options are ignored](https://github.com/openxla/xla/blob/b4abe20e7301e98abf641a7d65e0226c4247242b/xla/pjrt/pjrt_c_api_client.cc#L408). In practice, this means that JAX compilation cache fails when deserializing executables for rank zero on non-zero ranks.

JAX repro:

```python
import jax
import jax.numpy as jnp
import logging
import argparse
import socket

parser = argparse.ArgumentParser(
                    prog='mock-test',
                    description='Tests mocking',
                    epilog='...')
parser.add_argument('-r', '--rank', type=int)
args = parser.parse_args()

if args.rank == 1:
  logging.basicConfig(format='%(asctime)s %(message)s')
  logging.getLogger("jax._src.compiler").setLevel(logging.DEBUG)

jax.config.update("jax_compilation_cache_dir", "/tmp/compilation_cache")
jax.config.update("jax_persistent_cache_min_compile_time_secs", 0)

jax.distributed.initialize(socket.gethostname() + ":1234", 2, args.rank, local_device_ids = [args.rank])

f = jax.jit(lambda x: x)
jax.block_until_ready(f(jnp.zeros((1,))))
```

Run `(python test.py -r0 &) && (python test.py -r1)` twice on a machine with 2+ GPUs.

The output (from rank 1) then contains the following error message:
```
/opt/jax/jax/_src/compiler.py:691: UserWarning: Error reading persistent compilation cache entry for 'jit__lambda_':
XlaRuntimeError: INVALID_ARGUMENT: Device assignment (Computations: 1 Replicas: 1
Computation 0: 0
) does not have any local devices.
  warnings.warn(
2024-10-14 09:16:30,222 PERSISTENT COMPILATION CACHE MISS for 'jit__lambda_' with key 'jit__lambda_-502ff86f0064419e429f73e9641f94cc3ab91a275910dec17b3ba6186556a297'
```

### PR DESCRIPTION
**Pull request by jaro-sevcik**
This enables JAX to supply different device assignment when deserializing single-device executables from compilation cache.

Fixes https://github.com/openxla/xla/issues/18286.

### CODE REVIEW

File: xla/pjrt/c/pjrt_c_api.h
Comment (changm) [inline xla/pjrt/c/pjrt_c_api.h:1579]: Is this the same compile options proto as in Compile()? 

https://github.com/openxla/xla/blob/b46d80b8b651ca230d5b797a128dbd3dbe30de5c/xla/pjrt/c/pjrt_c_api.h#L2106

If so, mind adding a comment explicitly stating it's the same?

File: xla/pjrt/c/pjrt_c_api.h
Comment (changm) [inline xla/pjrt/c/pjrt_c_api.h:1580]: Also, due to ABI compatibility, I think these two arguments have to go at the end of the struct. This part is fuzzy to me due to the last arg here being the output.

@skye - Can you confirm here please?

File: xla/pjrt/c/pjrt_c_api_wrapper_impl.cc
Comment (jaro-sevcik) [reply xla/pjrt/c/pjrt_c_api_wrapper_impl.cc:1573]: I believe 'nullopt' means that the options will be taken from the serialized executable (i.e., exactly the options that were used for compilation).

File: xla/pjrt/c/pjrt_c_api.h
Comment (jaro-sevcik) [reply xla/pjrt/c/pjrt_c_api.h:1582]: Done.

File: xla/pjrt/c/pjrt_c_api.h
Comment (jaro-sevcik) [reply xla/pjrt/c/pjrt_c_api.h:1584]: Reordered.

**Comment by jaro-sevcik** [comment]
@changm Addressed the comments. Could you take another look, please?

File: xla/pjrt/c/pjrt_c_api_wrapper_impl.cc
Comment (changm) [reply xla/pjrt/c/pjrt_c_api_wrapper_impl.cc:1573]: Ahh let me try again :)

I don't think you need to make this optional in line 1573 and check for null in line 1575. Just always declare an instance of options and ParseCompileOptions. ParseCompileOptions should handle a nullptr just fine.

Does that work? Thanks!

**Review by changm — APPROVED** [review]
LGTM assuming last comment that you can fix up please :). Thank you!

File: xla/pjrt/c/pjrt_c_api_wrapper_impl.cc
Comment (jaro-sevcik) [reply xla/pjrt/c/pjrt_c_api_wrapper_impl.cc:1573]: `ParseCompiledOptions` does not seem to handle `nullptr`. It does not even return optional, it just returns parsed options or a (generic) error on parse failure. Are you suggesting to turn parse failures into default compile options?

The `nullopt` value is semantically different from default compile options, so we should not mix them up.

Does this make sense or am I missing something?

Comment (jaro-sevcik) [reply xla/pjrt/c/pjrt_c_api_wrapper_impl.cc:1573]: To answer your original question ("Any particular reason it's optional here but not during Compile()?"):

The `optional` here allows to optionally override the compile options that are already serialized in the `serialized` argument.

The `nullopt` options for the `DeserializeExecutable` method means that the runtime should use the serialized compile options from the `serialized` argument to set some of the properties of the in-memory executable object (such as device assignment).

The non-nullopt options for the `DeserializeExecutable` method override the serialized compile options and when setting the properties of the in-memory executable object (device assignment).

See here: https://github.com/openxla/xla/blob/eef7ee50d0980848436f0b4f402cec8c5bf86f21/xla/pjrt/pjrt_stream_executor_client.cc#L3667

[ The options for the `Compile` method do not optionally override anything, they are always used to set the properties of the executable object - there is no need for the `nullopt` value. ]

**Comment by jaro-sevcik** [comment]
I am still unsure what you had in mind in your last comment. I have added some additional explanation on the comment.

File: xla/pjrt/c/pjrt_c_api.h
Comment (changm) [inline xla/pjrt/c/pjrt_c_api.h:1583]: nit: Rename this to overridden_serialized_compile_options here and everywhere else

File: xla/pjrt/pjrt_c_api_client_test.cc
Comment (changm) [inline xla/pjrt/pjrt_c_api_client_test.cc:197]: nit: "Check that the executable's compile options were overrideen with device id 1"

File: xla/pjrt/c/pjrt_c_api.h
Comment (jaro-sevcik) [reply xla/pjrt/c/pjrt_c_api.h]: Done.

File: xla/pjrt/pjrt_c_api_client_test.cc
Comment (jaro-sevcik) [reply xla/pjrt/pjrt_c_api_client_test.cc]: Done.

**Comment by jaro-sevcik** [comment]
Addressed the comments, could you take another look?

**Review by changm — APPROVED** [review]

**Comment by changm** [comment]
Thanks LGTM!

**Comment by changm** [comment]
Not sure it's too late, but I also think you have to bump up the minor version number here since you changed the API - https://github.com/openxla/xla/blob/9f24c04c94c3c65c70ed9eec6275cc8c6a963eb6/xla/pjrt/c/pjrt_c_api.h#L82

**Review by changm — APPROVED** [review]

**Comment by nouiz** [comment]
Any idea what is blocking the PR as it is approved? We don't have visibility on the issue.

**Comment by changm** [comment]
Hmm sorry for the annoying things. Looks like you have to update the CHANGELOG here - [xla/pjrt/c/CHANGELOG.md](https://github.com/openxla/xla/blob/975f007cfc6045ec692d74ef2b5b06d3384abd2a/xla/pjrt/c/CHANGELOG.md#L4)

**Comment by changm** [comment]
I think this PR still needs to be updated to add the new version in changelog file. After that @golechwierowicz should be shepherding this through.

**Review by changm — APPROVED** [review]

**Review by golechwierowicz — APPROVED** [review]

**Review by changm — APPROVED** [review]

**Comment by changm** [comment]
Hmm, I think everything went through but not sure why this PR still is open. @golechwierowicz Any idea or whose a reviewer with write access?

**Comment by golechwierowicz** [comment]
It's in :)

**Comment by loislo** [comment]
it seems the pr was the reason for a failure in one of the prod pipeline. It was rolled back.

**Comment by jaro-sevcik** [comment]
> it seems the pr was the reason for a failure in one of the prod pipeline. It was rolled back.

@loislo Could you share more details on what exactly has failed?

Also, could you please reopen https://github.com/openxla/xla/issues/18286? Since you reverted the fix, JAX compilation caching is still broken on multi-node configs.

**Comment by loislo** [comment]
done.

**Comment by loislo** [comment]
the deserialization process was failing in some cases.
The observed behavior was:
the size of the override was nonzero and big.
the proto parsing was failing.

**Pull request closed without merging**
